homeassistant/components/neptune_apex/config_flow.py

```diff
@@ -0,0 +1,88 @@
+"""Config flow for Neptune Apex integration."""
+import logging
+
+from pynepsys import Apex
+import voluptuous as vol
+
+from homeassistant import config_entries, core, exceptions
+from homeassistant.const import CONF_HOST, CONF_PASSWORD, CONF_USERNAME
+
+from .const import DOMAIN  # pylint:disable=unused-import
+
+_LOGGER = logging.getLogger(__name__)
+
+DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_HOST): str,
+        vol.Required(CONF_USERNAME): str,
+        vol.Required(CONF_PASSWORD): str,
+    }
+)
+
+
+class ApexHub:
+    """Handle authentication validation during config flow."""
+
+    def __init__(self, host):
+        """Initialize."""
+        self.host = host
+        self.apex = None
+
+    async def authenticate(self, username, password) -> bool:
+        """Test if we can authenticate with the host."""
+        self.apex = Apex(self.host, username, password)
+        result = await self.apex.validate_connection()
+        if result == "success":
+            return True
+        if result == "invalid_auth":
+            raise InvalidAuth
+        raise CannotConnect
+
+
+async def validate_input(hass: core.HomeAssistant, data):
+    """Validate the user input allows us to connect.
+
+    Data has the keys from DATA_SCHEMA with values provided by the user.
+    """
+    hub = ApexHub(data[CONF_HOST])
+
+    if not await hub.authenticate(data[CONF_USERNAME], data[CONF_PASSWORD]):
+        raise InvalidAuth
+
+    # Return info that you want to store in the config entry.
+    return {"title": "Neptune Apex"}
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Neptune Apex."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_POLL
+
+    async def async_step_user(self, user_input=None):
+        """Handle the initial step."""
+        errors = {}
+        if user_input is not None:
+            try:
+                info = await validate_input(self.hass, user_input)
+
+                return self.async_create_entry(title=info["title"], data=user_input)
```

Comment:
We need to guard for creating the same entry more than once, ie with the same host. If host is an ip address, which can change, we can not use our unique id api with the host as an id.

If we would have something unique and stable we could use our unique id api for config entries:
https://developers.home-assistant.io/docs/config_entries_config_flow_handler#unique-ids

If there's no good unique id we just have to check existing entries for the same host.

Reply:
OK, I've added a hook during validate_input that reads the serial # off the device and I think creates the unique id. Being honest, I'd designed this component to support a single Apex, but this is an improvement.